### PR TITLE
Kernel<-Fusion Step 1 - Convert ExprSort to StmtSort

### PR DESCRIPTION
**Pull request by csarofeen**
Expand ExprSort to be able to grab all statements, sorted correctly so we also can have Vals.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp [hunk 1]
```diff
@@ -31,6 +32,93 @@ void remove_visited(
...
+  void handle(Statement* stmt) {
+    if (stmt->isVal()) {
+      handle(stmt->as<Val>());
+    } else if (stmt->isExpr()) {
+      handle(stmt->as<Expr>());
+    } else {
+      TORCH_INTERNAL_ASSERT(
+          false, "IterVisitor could not detect type in next_dispatch.");
+    }
+  }
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: Is this overload necessary? Build and C++ tests seem fine without it.

File: torch/csrc/jit/codegen/cuda/partial_split_map.cpp [hunk 2]
```diff
@@ -39,7 +39,6 @@ void PartialSplitMap::build(Fusion* fusion) {
 }
 
 Val* PartialSplitMap::getStartOffset(IterDomain* root_domain) const {
-  TORCH_INTERNAL_ASSERT(!root_domain->isKirStmt());
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/partial_split_map.cpp:42]: Does this removal have anything to do with the sorting change?

File: torch/csrc/jit/codegen/cuda/iter_visitor.cpp (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/iter_visitor.cpp]: Likely unnecessary, removing.

File: torch/csrc/jit/codegen/cuda/partial_split_map.cpp (hunk 2 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/partial_split_map.cpp:42]: No, likely grabbed this change in this PR by mistake. Would have been done in a later PR anyways.

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**